File: _includes/sort-benchmark/algorithms/unshuffle.rs
fn merge_two_sorted(left: &[usize], right: &[usize]) -> Vec<usize> {
    let mut out = Vec::with_capacity(left.len() + right.len());
    let mut i = 0;
    let mut j = 0;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            out.push(left[i]);
            i += 1;
        } else {
            out.push(right[j]);
            j += 1;
        }
    }
    out.extend_from_slice(&left[i..]);
    out.extend_from_slice(&right[j..]);
    out
}

fn unshuffle_sort(a: &mut [usize]) {
    let mut piles: Vec<Vec<usize>> = Vec::new();

    for &x in a.iter() {
        let mut placed = false;
        for pile in piles.iter_mut() {
            let front = pile[0];
            let back = *pile.last().unwrap();
            if x <= front {
                pile.insert(0, x);
                placed = true;
                break;
            } else if x >= back {
                pile.push(x);
                placed = true;
                break;
            }
        }
        if !placed {
            piles.push(vec![x]);
        }
    }

    if piles.is_empty() {
        return;
    }

    let mut merged = piles[0].clone();
    for pile in piles.into_iter().skip(1) {
        merged = merge_two_sorted(&merged, &pile);
    }

    a.copy_from_slice(&merged);
}

File: _includes/sort-benchmark/algorithms/unshuffle.swift
func merge_two_sorted(_ left: [Int], _ right: [Int]) -> [Int] {
    var out = [Int]()
    out.reserveCapacity(left.count + right.count)
    var i = 0
    var j = 0
    while i < left.count && j < right.count {
        if left[i] <= right[j] {
            out.append(left[i])
            i += 1
        } else {
            out.append(right[j])
            j += 1
        }
    }
    out.append(contentsOf: left[i...])
    out.append(contentsOf: right[j...])
    return out
}

func unshuffle_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { unshuffle_sort($0) }
}

func unshuffle_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    var piles = [[Int]]()

    for i in 0..<a.count {
        let x = a[i]
        var placed = false
        for p in 0..<piles.count {
            let front = piles[p][0]
            let back = piles[p].last!
            if x <= front {
                piles[p].insert(x, at: 0)
                placed = true
                break
            } else if x >= back {
                piles[p].append(x)
                placed = true
                break
            }
        }
        if !placed {
            piles.append([x])
        }
    }

    if piles.isEmpty {
        return
    }

    var merged = piles[0]
    for p in 1..<piles.count {
        merged = merge_two_sorted(merged, piles[p])
    }

    for i in 0..<a.count {
        a[i] = merged[i]
    }
}
